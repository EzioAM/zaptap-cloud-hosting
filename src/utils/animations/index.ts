/**
 * Animation System Export Hub
 * Central export point for all animation utilities
 */

// Core exports
export {
  AnimationController,
  animationController,
  useAnimationController,
  AnimationPresets,
  DevicePerformance,
} from './AnimationController';

// Platform optimizations
export {
  IOSOptimizations,
  AndroidOptimizations,
  WebOptimizations,
  PlatformOptimizer,
  AnimationPool,
  animationPool,
} from './PlatformOptimizations';

// Performance hooks
export {
  useAnimationPerformance,
  useOptimizedAnimatedValue,
  useDelayedAnimation,
  useBatchAnimations,
  useOptimizedScrollAnimation,
  useSpringAnimation,
  useTimingAnimation,
  useReducedMotion,
  useFPSMonitor,
  useLazyAnimation,
  useGestureAnimation,
} from './PerformanceHooks';

// Preset animations
export { PresetAnimations } from './PresetAnimations';

// Constants and configurations
export { PERFORMANCE_TARGETS, DURATIONS, SPRING_CONFIGS, Easings } from './constants';

// Animation helpers
import { Animated } from 'react-native';
import { animationController } from './AnimationController';

// Quick animation creators
export const fadeIn = (value: Animated.Value, duration: number = 300) => {
  return animationController.createTiming(value, 1, { duration });
};

export const fadeOut = (value: Animated.Value, duration: number = 300) => {
  return animationController.createTiming(value, 0, { duration });
};

export const slideIn = (value: Animated.Value, from: number = 100, duration: number = 300) => {
  value.setValue(from);
  return animationController.createSpring(value, 0);
};

export const slideOut = (value: Animated.Value, to: number = 100, duration: number = 300) => {
  return animationController.createSpring(value, to);
};

export const scaleIn = (value: Animated.Value, duration: number = 300) => {
  value.setValue(0);
  return animationController.createSpring(value, 1);
};

export const scaleOut = (value: Animated.Value, duration: number = 300) => {
  return animationController.createSpring(value, 0);
};