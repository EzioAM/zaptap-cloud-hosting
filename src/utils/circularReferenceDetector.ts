/**
 * Circular Reference Detection and Prevention Utilities
 * Prevents stack overflow errors from circular object references
 */

/**
 * Check if an object contains circular references
 */
export function hasCircularReference(obj: any): boolean {
  const seen = new WeakSet();
  
  function detect(current: any, depth: number = 0): boolean {
    // Prevent deep recursion
    if (depth > 100) {
      console.warn('Deep object detected, stopping at depth 100');
      return false;
    }
    
    if (current && typeof current === 'object') {
      if (seen.has(current)) {
        return true; // Circular reference found
      }
      seen.add(current);
      
      try {
        // Check arrays
        if (Array.isArray(current)) {
          for (let i = 0; i < current.length; i++) {
            if (detect(current[i], depth + 1)) return true;
          }
        } else {
          // Check object properties
          for (const key in current) {
            if (current.hasOwnProperty(key)) {
              if (detect(current[key], depth + 1)) return true;
            }
          }
        }
      } catch (error) {
        // Silently handle any errors (e.g., getter errors)
        return false;
      }
    }
    return false;
  }
  
  try {
    return detect(obj);
  } catch (error) {
    console.error('Error detecting circular reference:', error);
    return false;
  }
}

/**
 * Remove circular references from an object
 */
export function removeCircularReferences(obj: any): any {
  try {
    const seen = new WeakSet();
    
    return JSON.parse(JSON.stringify(obj, (key, value) => {
      if (typeof value === 'object' && value !== null) {
        if (seen.has(value)) {
          return '[Circular Reference Removed]';
        }
        seen.add(value);
      }
      return value;
    }));
  } catch (error) {
    console.error('Error removing circular references:', error);
    return null;
  }
}

/**
 * Safely stringify an object, handling circular references
 */
export function safeStringify(obj: any, space?: number): string {
  try {
    const seen = new WeakSet();
    return JSON.stringify(obj, (key, value) => {
      if (typeof value === 'object' && value !== null) {
        if (seen.has(value)) {
          return '[Circular]';
        }
        seen.add(value);
      }
      return value;
    }, space);
  } catch (error) {
    return '{"error": "Failed to stringify object"}';
  }
}

/**
 * Create a deep clone of an object without circular references
 */
export function safeClone<T>(obj: T): T {
  try {
    if (obj === null || typeof obj !== 'object') {
      return obj;
    }
    
    // Handle special cases
    if (obj instanceof Date) {
      return new Date(obj.getTime()) as any;
    }
    
    if (obj instanceof Array) {
      return obj.map(item => safeClone(item)) as any;
    }
    
    // Regular object - use JSON parse/stringify to break circular refs
    return removeCircularReferences(obj);
  } catch (error) {
    console.error('Error cloning object:', error);
    return {} as T;
  }
}

/**
 * Check if an object is safe to process (no circular refs, reasonable size)
 */
export function isSafeObject(obj: any, maxDepth: number = 50): boolean {
  if (obj === null || typeof obj !== 'object') {
    return true;
  }
  
  // Quick size check
  try {
    const jsonStr = JSON.stringify(obj);
    if (jsonStr.length > 1000000) { // 1MB limit
      console.warn('Object too large for safe processing');
      return false;
    }
  } catch (error) {
    // Likely has circular references
    return false;
  }
  
  return !hasCircularReference(obj);
}