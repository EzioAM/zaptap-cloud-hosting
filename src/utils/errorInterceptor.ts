/**
import { EventLogger } from './EventLogger';
 * Smart Error Interceptor
 * 
 * This module intelligently manages console.error calls to reduce spam
 * while maintaining visibility into real issues.
 */

// Keep reference to original console.error
const originalConsoleError = console.error;

// Track errors by type and context
interface ErrorInfo {
  count: number;
  firstSeen: number;
  lastSeen: number;
  endpoint?: string;
  suppressed: number;
}

const errorTracker = new Map<string, ErrorInfo>();

// Configuration
const ERROR_BATCH_WINDOW = 10000; // 10 seconds
const MAX_ERRORS_BEFORE_SUMMARY = 3; // Show first 3, then summarize
const SUMMARY_INTERVAL = 30000; // Show summary every 30 seconds

// Last summary time
let lastSummaryTime = 0;

/**
 * Network error patterns
 */
const NETWORK_ERROR_PATTERNS = [
  'Network request failed',
  'Failed to fetch',
  'NetworkError',
  'AuthRetryableFetchError',
  'fetch failed',
  'ERR_NETWORK',
  'ERR_INTERNET_DISCONNECTED',
  'ECONNREFUSED',
  'ETIMEDOUT',
  'Unable to resolve host',
  'Network is unreachable',
];

/**
 * Extract error context (like API endpoint or component)
 */
const extractErrorContext = (args: any[]): { key: string; endpoint?: string } => {
  let key = 'unknown_error';
  let endpoint: string | undefined;

  // Look for API endpoint information
  for (const arg of args) {
    if (typeof arg === 'string') {
      // Check for common API patterns
      if (arg.includes('getPublicAutomations')) {
        key = 'api_public_automations';
        endpoint = 'getPublicAutomations';
      } else if (arg.includes('getTrendingAutomations')) {
        key = 'api_trending_automations';
        endpoint = 'getTrendingAutomations';
      } else if (arg.includes('auth') || arg.includes('session')) {
        key = 'auth_session';
        endpoint = 'auth';
      } else if (arg.includes('RPC function')) {
        const match = arg.match(/RPC function (\w+)/);
        if (match) {
          key = `rpc_${match[1]}`;
          endpoint = match[1];
        }
      }
    }
    
    // Check error objects
    if (arg && typeof arg === 'object' && arg.message) {
      if (arg.message.includes('Network request failed')) {
        if (key === 'unknown_error') {
          key = 'network_request_failed';
        }
      }
    }
  }

  return { key, endpoint };
};

/**
 * Check if error is network-related
 */
const isNetworkError = (args: any[]): boolean => {
  for (const arg of args) {
    if (arg instanceof Error && arg.name === 'TypeError' && 
        arg.message && arg.message.includes('Network request failed')) {
      return true;
    }
    
    if (arg && typeof arg === 'object') {
      if (arg.name === 'NetworkError' || 
          arg.name === 'AuthRetryableFetchError' ||
          (arg.message && NETWORK_ERROR_PATTERNS.some(pattern => arg.message.includes(pattern)))) {
        return true;
      }
    }
  }
  
  const errorString = args.map(arg => String(arg)).join(' ');
  return NETWORK_ERROR_PATTERNS.some(pattern => errorString.includes(pattern));
};

/**
 * Show error summary
 */
const showErrorSummary = () => {
  const now = Date.now();
  if (now - lastSummaryTime < SUMMARY_INTERVAL) {
    return;
  }
  
  lastSummaryTime = now;
  
  const networkErrors = Array.from(errorTracker.entries())
    .filter(([key]) => key.includes('api_') || key.includes('rpc_') || key.includes('network'))
    .filter(([, info]) => info.suppressed > 0);

  if (networkErrors.length > 0) {
    EventLogger.debug('errorInterceptor', '\nðŸ“Š Network Error Summary (last 30s):');
    networkErrors.forEach(([key, info]) => {
      const total = info.count + info.suppressed;
      const endpoint = info.endpoint || key;
      EventLogger.debug('errorInterceptor', '   - ${endpoint}: ${total} errors (${info.count} shown, ${info.suppressed} suppressed)');
    });
    EventLogger.debug('errorInterceptor', 'ðŸ’¡ Check your network connection or API availability\n');
  }
};

/**
 * Initialize the error interceptor
 */
export const initializeErrorInterceptor = () => {
  console.error = (...args: any[]) => {
    const now = Date.now();
    const { key, endpoint } = extractErrorContext(args);
    
    // Get or create error info
    let errorInfo = errorTracker.get(key);
    if (!errorInfo) {
      errorInfo = {
        count: 0,
        firstSeen: now,
        lastSeen: now,
        endpoint,
        suppressed: 0,
      };
      errorTracker.set(key, errorInfo);
    }
    
    // Update last seen
    errorInfo.lastSeen = now;
    
    // Check if this is a network error
    if (isNetworkError(args)) {
      // For network errors, show first few then suppress
      if (errorInfo.count < MAX_ERRORS_BEFORE_SUMMARY) {
        errorInfo.count++;
        // Show with context
        const context = endpoint ? ` [${endpoint}]` : '';
        EventLogger.debug('errorInterceptor', 'ðŸ”´ Network Error${context}:', args[0]?.message || args[0] || 'Network request failed');
      } else {
        errorInfo.suppressed++;
        
        // Show summary periodically
        if (errorInfo.suppressed % 10 === 0) {
          EventLogger.debug('errorInterceptor', 'ðŸ“´ Network errors continuing for ${endpoint || key} (${errorInfo.suppressed + errorInfo.count} total)');
        }
      }
      
      // Show summary if enough time has passed
      showErrorSummary();
      return;
    }
    
    // For non-network errors, always show
    originalConsoleError.apply(console, args);
  };
  
  // Set up periodic summary
  setInterval(showErrorSummary, SUMMARY_INTERVAL);
};

/**
 * Restore original console.error
 */
export const restoreConsoleError = () => {
  console.error = originalConsoleError;
};

/**
 * Reset error tracking
 */
export const resetErrorTracking = () => {
  errorTracker.clear();
  lastSummaryTime = 0;
};

/**
 * Get error statistics
 */
export const getErrorStats = () => {
  const stats: Record<string, any> = {};
  errorTracker.forEach((info, key) => {
    stats[key] = {
      total: info.count + info.suppressed,
      shown: info.count,
      suppressed: info.suppressed,
      endpoint: info.endpoint,
    };
  });
  return stats;
};