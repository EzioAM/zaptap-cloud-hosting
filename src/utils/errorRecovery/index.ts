export { ErrorRecoveryManager, errorRecoveryManager } from './ErrorRecoveryManager';
export { RetryWrapper, useErrorBoundary, useErrorHandler } from './RetryWrapper';

export type { 
  RetryConfig, 
  RecoveryStrategy,
  RetryWrapperProps,
  RetryFallbackProps,
} from './ErrorRecoveryManager';