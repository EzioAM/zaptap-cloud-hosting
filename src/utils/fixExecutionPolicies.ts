import { supabase } from '../services/supabase/client';

export async function fixExecutionPolicies() {
  try {
    console.log('[FixPolicies] Starting to fix execution policies...');
    
    // First, let's check if the policies exist
    const { data: existingPolicies, error: checkError } = await supabase
      .rpc('get_policies_for_table', { table_name: 'automation_execution_summary' });
    
    if (checkError) {
      console.error('[FixPolicies] Error checking policies:', checkError);
    } else {
      console.log('[FixPolicies] Existing policies:', existingPolicies);
    }
    
    // Try to create the missing policies using raw SQL
    const queries = [
      `
      CREATE POLICY IF NOT EXISTS "System can insert execution summaries" 
      ON public.automation_execution_summary
      FOR INSERT WITH CHECK (
        EXISTS (
          SELECT 1 FROM public.automations 
          WHERE automations.id = automation_execution_summary.automation_id 
          AND automations.user_id = (select auth.uid())
        )
      );
      `,
      `
      CREATE POLICY IF NOT EXISTS "System can update execution summaries" 
      ON public.automation_execution_summary
      FOR UPDATE USING (
        EXISTS (
          SELECT 1 FROM public.automations 
          WHERE automations.id = automation_execution_summary.automation_id 
          AND automations.user_id = (select auth.uid())
        )
      );
      `,
      `
      DROP POLICY IF EXISTS "Users can update their execution records" ON public.automation_executions;
      CREATE POLICY "Users can update their execution records" ON public.automation_executions
      FOR UPDATE USING (
        user_id = (select auth.uid()) OR
        EXISTS (
          SELECT 1 FROM public.automations 
          WHERE automations.id = automation_executions.automation_id 
          AND automations.user_id = (select auth.uid())
        )
      );
      `
    ];
    
    for (const query of queries) {
      const { error } = await supabase.rpc('exec_sql', { sql_query: query });
      if (error) {
        console.error('[FixPolicies] Error executing query:', error);
      } else {
        console.log('[FixPolicies] Query executed successfully');
      }
    }
    
    console.log('[FixPolicies] Policies fix attempt completed');
  } catch (error) {
    console.error('[FixPolicies] Failed to fix policies:', error);
  }
}

// Alternative: Skip the summary table update entirely in AutomationEngine
export function shouldSkipSummaryUpdate(): boolean {
  // We can use this flag to temporarily skip summary updates
  // until the RLS policies are fixed in production
  return true; // Set to true to skip summary updates
}