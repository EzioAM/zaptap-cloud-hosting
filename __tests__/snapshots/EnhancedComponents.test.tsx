import React from 'react';
import renderer from 'react-test-renderer';
import { renderWithAllProviders } from '../utils/renderWithProviders';
import { TestDataFactory } from '../utils/testHelpers';

// Create mock components for snapshot testing since the actual enhanced components may not exist yet
const MockBuildScreenSafe = ({ automation }: { automation?: any }) => (
  <div data-testid="build-screen-safe">
    <div className="header">
      <h1>Build Automation</h1>
      <div className="breadcrumb">Home &gt; Build &gt; New</div>
    </div>
    
    <div className="form-section">
      <input 
        type="text" 
        placeholder="Automation Title"
        defaultValue={automation?.title || ''}
        className="title-input"
      />
      <textarea 
        placeholder="Description (optional)"
        defaultValue={automation?.description || ''}
        className="description-input"
      />
    </div>

    <div className="step-palette">
      <h3>Step Types</h3>
      <div className="step-categories">
        <div className="category">
          <h4>Communication</h4>
          <div className="step-item">📱 Send SMS</div>
          <div className="step-item">📧 Send Email</div>
          <div className="step-item">🔔 Notification</div>
        </div>
        
        <div className="category">
          <h4>Web & Data</h4>
          <div className="step-item">🌐 Webhook</div>
          <div className="step-item">📊 HTTP Request</div>
          <div className="step-item">💾 Store Data</div>
        </div>
        
        <div className="category">
          <h4>Device</h4>
          <div className="step-item">📸 Take Photo</div>
          <div className="step-item">📍 Get Location</div>
          <div className="step-item">🔊 Play Sound</div>
        </div>
      </div>
    </div>

    <div className="steps-container">
      <h3>Steps ({automation?.steps?.length || 0})</h3>
      {(!automation?.steps || automation.steps.length === 0) ? (
        <div className="empty-state">
          <div className="empty-icon">📋</div>
          <p>No steps added yet</p>
          <p className="subtitle">Add steps from the palette above</p>
        </div>
      ) : (
        <div className="steps-list">
          {automation.steps.map((step: any, index: number) => (
            <div key={step.id} className="step-card">
              <div className="step-number">{index + 1}</div>
              <div className="step-content">
                <h4>{step.title}</h4>
                <p className="step-type">Type: {step.type}</p>
              </div>
              <div className="step-actions">
                <button className="move-up" disabled={index === 0}>↑</button>
                <button className="move-down" disabled={index === automation.steps.length - 1}>↓</button>
                <button className="remove">×</button>
              </div>
            </div>
          ))}
        </div>
      )}
    </div>

    <div className="save-section">
      <button className="save-button primary">Save Automation</button>
      <button className="cancel-button secondary">Cancel</button>
    </div>
  </div>
);

const MockDeploymentOptions = ({ automation }: { automation: any }) => (
  <div data-testid="deployment-options" className="modal-overlay">
    <div className="modal-content">
      <div className="modal-header">
        <h2>Deploy Automation</h2>
        <button className="close-button">×</button>
      </div>
      
      <div className="automation-preview">
        <h3>{automation.title}</h3>
        <p className="automation-description">{automation.description}</p>
        <div className="automation-stats">
          <span className="step-count">{automation.steps?.length || 0} steps</span>
          <span className="visibility">{automation.is_public ? 'Public' : 'Private'}</span>
        </div>
      </div>

      <div className="deployment-methods">
        <div className="method-card nfc">
          <div className="method-icon">📱</div>
          <h4>NFC Tag</h4>
          <p>Write to NFC tag for instant execution</p>
          <button className="deploy-button">Write to NFC</button>
        </div>

        <div className="method-card qr">
          <div className="method-icon">🔍</div>
          <h4>QR Code</h4>
          <p>Generate QR code for easy sharing</p>
          <button className="deploy-button">Generate QR</button>
        </div>

        <div className="method-card share">
          <div className="method-icon">🔗</div>
          <h4>Share Link</h4>
          <p>Create shareable web link</p>
          <button className="deploy-button">Create Link</button>
        </div>
      </div>

      <div className="modal-footer">
        <button className="close-modal">Close</button>
      </div>
    </div>
  </div>
);

const MockOfflineIndicator = ({ isOnline }: { isOnline: boolean }) => (
  <div 
    data-testid="offline-indicator" 
    className={`offline-indicator ${!isOnline ? 'offline' : 'online'}`}
  >
    <div className="indicator-content">
      <div className="status-icon">
        {isOnline ? '🟢' : '🔴'}
      </div>
      <span className="status-text">
        {isOnline ? 'Online' : 'Offline'}
      </span>
      {!isOnline && (
        <div className="queue-info">
          <span className="queue-count">3 operations queued</span>
        </div>
      )}
    </div>
  </div>
);

const MockOnboardingFlow = ({ currentStep }: { currentStep: number }) => {
  const steps = [
    { title: 'Welcome', subtitle: 'Welcome to ZapTap' },
    { title: 'Permissions', subtitle: 'Grant required permissions' },
    { title: 'NFC Setup', subtitle: 'Learn about NFC tags' },
    { title: 'First Automation', subtitle: 'Create your first automation' },
  ];

  const currentStepData = steps[currentStep] || steps[0];

  return (
    <div data-testid="onboarding-flow" className="onboarding-container">
      <div className="onboarding-header">
        <div className="progress-indicator">
          Step {currentStep + 1} of {steps.length}
        </div>
        <div className="progress-bar">
          <div 
            className="progress-fill"
            style={{ width: `${((currentStep + 1) / steps.length) * 100}%` }}
          />
        </div>
      </div>

      <div className="onboarding-content">
        <div className="step-illustration">
          {currentStep === 0 && '🎉'}
          {currentStep === 1 && '🔐'}
          {currentStep === 2 && '📱'}
          {currentStep === 3 && '⚡'}
        </div>
        
        <h1>{currentStepData.title}</h1>
        <p className="step-subtitle">{currentStepData.subtitle}</p>

        {currentStep === 0 && (
          <div className="welcome-features">
            <div className="feature-item">
              <span className="feature-icon">📱</span>
              <span>NFC Tag Automation</span>
            </div>
            <div className="feature-item">
              <span className="feature-icon">🔍</span>
              <span>QR Code Sharing</span>
            </div>
            <div className="feature-item">
              <span className="feature-icon">🔄</span>
              <span>Offline Support</span>
            </div>
          </div>
        )}

        {currentStep === 1 && (
          <div className="permissions-list">
            <div className="permission-item">
              <span className="permission-icon">🔔</span>
              <span>Notifications</span>
              <button className="grant-button">Grant</button>
            </div>
            <div className="permission-item">
              <span className="permission-icon">📷</span>
              <span>Camera</span>
              <button className="grant-button">Grant</button>
            </div>
            <div className="permission-item">
              <span className="permission-icon">📍</span>
              <span>Location</span>
              <button className="grant-button optional">Optional</button>
            </div>
          </div>
        )}

        {currentStep === 2 && (
          <div className="nfc-tutorial">
            <div className="tutorial-steps">
              <div className="tutorial-step">
                <span className="step-number">1</span>
                <span>Create an automation</span>
              </div>
              <div className="tutorial-step">
                <span className="step-number">2</span>
                <span>Tap "Deploy to NFC"</span>
              </div>
              <div className="tutorial-step">
                <span className="step-number">3</span>
                <span>Hold NFC tag near device</span>
              </div>
              <div className="tutorial-step">
                <span className="step-number">4</span>
                <span>Share or use your NFC tag!</span>
              </div>
            </div>
          </div>
        )}

        {currentStep === 3 && (
          <div className="automation-tutorial">
            <div className="tutorial-preview">
              <div className="mock-automation">
                <h4>Sample Automation</h4>
                <div className="sample-steps">
                  <div className="sample-step">📱 Send SMS to contact</div>
                  <div className="sample-step">🔔 Show notification</div>
                  <div className="sample-step">📧 Send email report</div>
                </div>
              </div>
            </div>
          </div>
        )}
      </div>

      <div className="onboarding-footer">
        {currentStep > 0 && (
          <button className="back-button">Back</button>
        )}
        
        <div className="navigation-buttons">
          {currentStep < steps.length - 1 && (
            <button className="skip-button">Skip</button>
          )}
          <button className="next-button primary">
            {currentStep === steps.length - 1 ? 'Get Started' : 'Next'}
          </button>
        </div>
      </div>
    </div>
  );
};

describe('Enhanced Components Snapshots', () => {
  describe('BuildScreenSafe', () => {
    it('should render empty build screen correctly', () => {
      const tree = renderer
        .create(<MockBuildScreenSafe />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render build screen with automation data', () => {
      const automation = TestDataFactory.createMockAutomation({
        title: 'Test Automation',
        description: 'A test automation for snapshot testing',
        steps: [
          {
            id: '1',
            type: 'sms',
            title: 'Send SMS',
            config: { message: 'Test message' },
          },
          {
            id: '2',
            type: 'notification',
            title: 'Show Notification',
            config: { title: 'Test notification' },
          },
        ],
      });

      const tree = renderer
        .create(<MockBuildScreenSafe automation={automation} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render build screen with many steps', () => {
      const automation = TestDataFactory.createMockAutomation({
        title: 'Complex Automation',
        description: 'An automation with many steps for testing layout',
        steps: [
          { id: '1', type: 'sms', title: 'Send SMS', config: {} },
          { id: '2', type: 'email', title: 'Send Email', config: {} },
          { id: '3', type: 'notification', title: 'Show Notification', config: {} },
          { id: '4', type: 'webhook', title: 'Call Webhook', config: {} },
          { id: '5', type: 'location', title: 'Get Location', config: {} },
        ],
      });

      const tree = renderer
        .create(<MockBuildScreenSafe automation={automation} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });
  });

  describe('DeploymentOptions', () => {
    it('should render deployment modal for public automation', () => {
      const automation = TestDataFactory.createMockAutomation({
        title: 'Public Test Automation',
        description: 'A public automation for deployment testing',
        is_public: true,
        steps: [
          { id: '1', type: 'sms', title: 'Send SMS', config: {} },
          { id: '2', type: 'notification', title: 'Show Notification', config: {} },
        ],
      });

      const tree = renderer
        .create(<MockDeploymentOptions automation={automation} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render deployment modal for private automation', () => {
      const automation = TestDataFactory.createMockAutomation({
        title: 'Private Test Automation',
        description: 'A private automation for deployment testing',
        is_public: false,
        steps: [
          { id: '1', type: 'email', title: 'Send Email', config: {} },
        ],
      });

      const tree = renderer
        .create(<MockDeploymentOptions automation={automation} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render deployment modal with no description', () => {
      const automation = TestDataFactory.createMockAutomation({
        title: 'No Description Automation',
        description: '',
        steps: [],
      });

      const tree = renderer
        .create(<MockDeploymentOptions automation={automation} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });
  });

  describe('OfflineIndicator', () => {
    it('should render online state correctly', () => {
      const tree = renderer
        .create(<MockOfflineIndicator isOnline={true} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render offline state correctly', () => {
      const tree = renderer
        .create(<MockOfflineIndicator isOnline={false} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });
  });

  describe('OnboardingFlow', () => {
    it('should render welcome step correctly', () => {
      const tree = renderer
        .create(<MockOnboardingFlow currentStep={0} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render permissions step correctly', () => {
      const tree = renderer
        .create(<MockOnboardingFlow currentStep={1} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render NFC tutorial step correctly', () => {
      const tree = renderer
        .create(<MockOnboardingFlow currentStep={2} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });

    it('should render automation tutorial step correctly', () => {
      const tree = renderer
        .create(<MockOnboardingFlow currentStep={3} />)
        .toJSON();
      
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Responsive Design Snapshots', () => {
    it('should render components with different screen sizes', () => {
      // This would typically involve setting different viewport sizes
      // For now, we'll just test the default rendering
      const automation = TestDataFactory.createMockAutomation();

      const buildScreen = renderer
        .create(<MockBuildScreenSafe automation={automation} />)
        .toJSON();

      const deploymentOptions = renderer
        .create(<MockDeploymentOptions automation={automation} />)
        .toJSON();

      expect(buildScreen).toMatchSnapshot('build-screen-responsive');
      expect(deploymentOptions).toMatchSnapshot('deployment-options-responsive');
    });
  });

  describe('Error States', () => {
    it('should render components with error states', () => {
      // Mock error states
      const ErrorBuildScreen = () => (
        <div className="build-screen error-state">
          <div className="error-banner">
            <span className="error-icon">⚠️</span>
            <span>Failed to load automation data</span>
            <button className="retry-button">Retry</button>
          </div>
          <MockBuildScreenSafe />
        </div>
      );

      const ErrorDeploymentOptions = () => (
        <div className="deployment-options error-state">
          <div className="error-message">
            <span className="error-icon">❌</span>
            <span>Deployment failed. Please try again.</span>
          </div>
          <MockDeploymentOptions automation={TestDataFactory.createMockAutomation()} />
        </div>
      );

      const buildTree = renderer.create(<ErrorBuildScreen />).toJSON();
      const deployTree = renderer.create(<ErrorDeploymentOptions />).toJSON();

      expect(buildTree).toMatchSnapshot('build-screen-error');
      expect(deployTree).toMatchSnapshot('deployment-options-error');
    });
  });

  describe('Loading States', () => {
    it('should render components with loading states', () => {
      const LoadingBuildScreen = () => (
        <div className="build-screen loading-state">
          <div className="loading-overlay">
            <div className="spinner"></div>
            <span>Loading automation...</span>
          </div>
          <MockBuildScreenSafe />
        </div>
      );

      const LoadingDeploymentOptions = () => (
        <div className="deployment-options loading-state">
          <div className="loading-banner">
            <div className="loading-spinner"></div>
            <span>Preparing deployment...</span>
          </div>
          <MockDeploymentOptions automation={TestDataFactory.createMockAutomation()} />
        </div>
      );

      const buildTree = renderer.create(<LoadingBuildScreen />).toJSON();
      const deployTree = renderer.create(<LoadingDeploymentOptions />).toJSON();

      expect(buildTree).toMatchSnapshot('build-screen-loading');
      expect(deployTree).toMatchSnapshot('deployment-options-loading');
    });
  });

  describe('Theme Variations', () => {
    it('should render components with dark theme', () => {
      const DarkThemeBuildScreen = () => (
        <div className="build-screen dark-theme">
          <MockBuildScreenSafe automation={TestDataFactory.createMockAutomation()} />
        </div>
      );

      const DarkThemeOnboarding = () => (
        <div className="onboarding-flow dark-theme">
          <MockOnboardingFlow currentStep={0} />
        </div>
      );

      const buildTree = renderer.create(<DarkThemeBuildScreen />).toJSON();
      const onboardingTree = renderer.create(<DarkThemeOnboarding />).toJSON();

      expect(buildTree).toMatchSnapshot('build-screen-dark-theme');
      expect(onboardingTree).toMatchSnapshot('onboarding-dark-theme');
    });

    it('should render components with light theme', () => {
      const LightThemeBuildScreen = () => (
        <div className="build-screen light-theme">
          <MockBuildScreenSafe automation={TestDataFactory.createMockAutomation()} />
        </div>
      );

      const LightThemeOfflineIndicator = () => (
        <div className="offline-indicator light-theme">
          <MockOfflineIndicator isOnline={false} />
        </div>
      );

      const buildTree = renderer.create(<LightThemeBuildScreen />).toJSON();
      const indicatorTree = renderer.create(<LightThemeOfflineIndicator />).toJSON();

      expect(buildTree).toMatchSnapshot('build-screen-light-theme');
      expect(indicatorTree).toMatchSnapshot('offline-indicator-light-theme');
    });
  });

  describe('Accessibility Snapshots', () => {
    it('should render components with accessibility attributes', () => {
      const AccessibleBuildScreen = () => (
        <div 
          className="build-screen" 
          role="main" 
          aria-label="Automation Builder"
        >
          <MockBuildScreenSafe automation={TestDataFactory.createMockAutomation()} />
        </div>
      );

      const AccessibleOnboarding = () => (
        <div 
          className="onboarding-flow" 
          role="dialog" 
          aria-labelledby="onboarding-title"
          aria-describedby="onboarding-description"
        >
          <MockOnboardingFlow currentStep={1} />
        </div>
      );

      const buildTree = renderer.create(<AccessibleBuildScreen />).toJSON();
      const onboardingTree = renderer.create(<AccessibleOnboarding />).toJSON();

      expect(buildTree).toMatchSnapshot('build-screen-accessible');
      expect(onboardingTree).toMatchSnapshot('onboarding-accessible');
    });
  });
});