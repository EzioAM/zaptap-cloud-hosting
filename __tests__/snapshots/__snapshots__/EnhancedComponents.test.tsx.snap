// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enhanced Components Snapshots Accessibility Snapshots should render components with accessibility attributes: build-screen-accessible 1`] = `null`;

exports[`Enhanced Components Snapshots Accessibility Snapshots should render components with accessibility attributes: onboarding-accessible 1`] = `null`;

exports[`Enhanced Components Snapshots BuildScreenSafe should render build screen with automation data 1`] = `null`;

exports[`Enhanced Components Snapshots BuildScreenSafe should render build screen with many steps 1`] = `null`;

exports[`Enhanced Components Snapshots BuildScreenSafe should render empty build screen correctly 1`] = `null`;

exports[`Enhanced Components Snapshots DeploymentOptions should render deployment modal for private automation 1`] = `null`;

exports[`Enhanced Components Snapshots DeploymentOptions should render deployment modal for public automation 1`] = `null`;

exports[`Enhanced Components Snapshots DeploymentOptions should render deployment modal with no description 1`] = `null`;

exports[`Enhanced Components Snapshots Error States should render components with error states: build-screen-error 1`] = `null`;

exports[`Enhanced Components Snapshots Error States should render components with error states: deployment-options-error 1`] = `null`;

exports[`Enhanced Components Snapshots Loading States should render components with loading states: build-screen-loading 1`] = `null`;

exports[`Enhanced Components Snapshots Loading States should render components with loading states: deployment-options-loading 1`] = `null`;

exports[`Enhanced Components Snapshots OfflineIndicator should render offline state correctly 1`] = `null`;

exports[`Enhanced Components Snapshots OfflineIndicator should render online state correctly 1`] = `null`;

exports[`Enhanced Components Snapshots OnboardingFlow should render NFC tutorial step correctly 1`] = `null`;

exports[`Enhanced Components Snapshots OnboardingFlow should render automation tutorial step correctly 1`] = `null`;

exports[`Enhanced Components Snapshots OnboardingFlow should render permissions step correctly 1`] = `null`;

exports[`Enhanced Components Snapshots OnboardingFlow should render welcome step correctly 1`] = `null`;

exports[`Enhanced Components Snapshots Responsive Design Snapshots should render components with different screen sizes: build-screen-responsive 1`] = `null`;

exports[`Enhanced Components Snapshots Responsive Design Snapshots should render components with different screen sizes: deployment-options-responsive 1`] = `null`;

exports[`Enhanced Components Snapshots Theme Variations should render components with dark theme: build-screen-dark-theme 1`] = `null`;

exports[`Enhanced Components Snapshots Theme Variations should render components with dark theme: onboarding-dark-theme 1`] = `null`;

exports[`Enhanced Components Snapshots Theme Variations should render components with light theme: build-screen-light-theme 1`] = `null`;

exports[`Enhanced Components Snapshots Theme Variations should render components with light theme: offline-indicator-light-theme 1`] = `null`;
